drivers: adc: sam0: Fix driver and sample

attie-argentum opened the pull request:
The `adc_sam0` driver previously expected `sequence->channels == 1`, which is not compatible with the ADC sample, and might be unexpected. In an attempt to avoid breaking changes, I have revised this to require a single bit set (with no confirmation as to whether it is the correct bit). The input channel for this driver needs to be set using the "_configurable inputs_" API instead - differential or not.

I have also revised `adc_sam0` so that it supports devices with `MCLK`, but only one ADC. This will be important for SAML21 support (see #41308).

Finally, I have revised the `drivers/adc` sample so that it will operate the "_configurable inputs_" API correctly.

This PR is split out from #41308.

attie-argentum committed:
drivers: adc: adc_sam0: Fix interpretation of channels

Previously this was expected to be equal to 1 at all times. This doesn't
play well with the sample or other users (e.g: adc_shell). Instead, we
should count the number of active channels in the bitfield, and ensure
that only one is identified.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum committed:
samples: drivers: adc: Revise handling of configurable inputs

All drivers that require input_positive to be set should be given it.
Some parts may require an offset, so abstract this away too.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -231,20 +231,27 @@ static int adc_sam0_channel_setup(const struct device *dev,
+#else
+		return -EINVAL;
 #endif
 	} else {
 		INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND;
 
```
why not change `INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND` to use ADC_INPUTCTRL_MUXNEG(value) ?

attie-argentum replied:
```
INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND;
```

would become

```
INPUTCTRL |= ADC_INPUTCTRL_MUXNEG(ADC_INPUTCTRL_MUXNEG_GND_Val);
```

... I see no particular benefit, and this isn't modified in the patch anyway. Can do if that's what you want though(?)

nandojve replied:
I look the macro code and it doesn't clear regs, so, there is no point in use it, you are right. My concern is that, let's assume that driver set `INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_AIN3` first. What happen if you add `INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND` ?

Maybe driver should first clear all channels and then set the new one.

```c
INPUTCTRL &= ~ADC_INPUTCTRL_MUXNEG_Msk;
INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND;
```

attie-argentum replied:
Note that `INPUTCTRL` here is a local variable, not a register... I think this is fine.

https://github.com/zephyrproject-rtos/zephyr/blob/0d412c4ce75c2f2ae0a6eb538627bbd9b04278ed/drivers/adc/adc_sam0.c#L185

attie-argentum replied:
@nandojve please confirm you're okay with this staying?

nandojve replied:
> Note that INPUTCTRL here is a local variable, not a register... I think this is fine.

uhmm, upper case is not for local variables. Please, fix it!

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -367,22 +380,44 @@ static int start_read(const struct device *dev,
+		}
+#else
+		return -EINVAL;
+#endif
 
 		break;
```
Drop empty line

attie-argentum replied:
Again, not really part of the patch, but can do if you want.

nandojve replied:
I prefer keep consistency in the code. Please, drop line 421.

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -231,20 +231,27 @@ static int adc_sam0_channel_setup(const struct device *dev,
 
-#ifndef ADC_INPUTCTRL_DIFFMODE
+#if defined(ADC_INPUTCTRL_DIFFMODE)
+		/* no op */
```
Why?

attie-argentum replied:
Why the `/* no op */`?

Because this makes it explicit that there are three paths available here (not just two), and also maintains symmetry with the other side of the `if {} else {}`.

Can drop it if you want.

nandojve replied:
Driver set this:
`INPUTCTRL |= ADC_INPUTCTRL_DIFFMODE;`
then
`INPUTCTRL |= ADC_INPUTCTRL_MUXNEG_GND;`

So, when driver set `ADC_INPUTCTRL_MUXNEG_GND` it automatically disable `ADC_INPUTCTRL_DIFFMODE` ?

attie-argentum replied:
> So, when driver set ADC_INPUTCTRL_MUXNEG_GND it automatically disable ADC_INPUTCTRL_DIFFMODE ?

By virtue of the `int INPUTCTRL = 0` above, yes...?

This "_collected_" value is finally written to the register here:

https://github.com/zephyrproject-rtos/zephyr/blob/0d412c4ce75c2f2ae0a6eb538627bbd9b04278ed/drivers/adc/adc_sam0.c#L235

---

`DIFFMODE` and `DSEQSTOP` fields are not present in this register for the SAML21 or SAMD21(resolved in `fixup_adc.h`)... snippet below is from SAMD5x.

![image](https://user-images.githubusercontent.com/36298367/146964413-91791745-518e-4a75-881a-3f000ef103d0.png)

attie-argentum replied:
@nandojve please resolve if you're happy, or let me know if not?

nandojve reviewed (changes requested):
Thank you for this PR @attie-argentum ,

I particulary don't like `#if .. #endif` in code but sometimes we need to use. Did you considered add a fixup for adc? It could be something similar to those at soc/arm/atmel_sam0/common. https://github.com/zephyrproject-rtos/zephyr/blob/f0765409a72af21328600d0d67b876439b4ca539/soc/arm/atmel_sam0/common/sercom_fixup_samd5x.h#L7

attie-argentum commented on drivers/adc/adc_sam0.c:
```diff
@@ -392,7 +393,20 @@ static int start_read(const struct device *dev,
+		}
+	}
+	if (active_channels != 1U) {
+		/* The caller is expected to identify a single input channel, which will
+		 * typically be the positive input, though no check is made for this...
+		 *
+		 * While ensuring that the channels bitfield matches the positive input
+		 * might be sensible, this will likely break users before this revision
+		 * was put in place.
+		 */
```
I make that first line (403) equal to 80 cols...?

attie-argentum commented:
@nandojve thanks for your comments.

I didn't consider a `fixup.h`, as I didn't know they existed... I'm not convinced it's the correct approach here, as we'd need to set it up for all supported parts... this way the driver targets the three known implementations (i.e: SAML2x, SAMD2x, SAME5x). If there become more possibilities, then it might become sensible, but in the interest of keeping changes small to avoid breaking things, I think I'd pick this route.

Regarding your 4x remaining comments, I hope my reasoning stands. If not, I'll do what I think you're after...

attie-argentum committed:
samples: drivers: adc: Add a device tree overlay for atsamd21_xpro

Add a basic device tree overlay to support the use of the ATSAMD21 XPro
dev board with the ADC sample application.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented:
@nandojve I believe I have now implemented an `adc_fixup.h` as requested. I've left this as a separate patch, as it's quite far-reaching...

nandojve commented on soc/arm/atmel_sam0/same51/soc.h:
```diff
@@ -32,6 +32,8 @@
 
 #endif /* _ASMLANGUAGE */
 
+#include "adc_fixup.h"
+
 #include "sercom_fixup_samd5x.h"
```
I will suggest to change fixup name to `adc_fixup_sam0.h` This make explicit this is only related to sam0 platform.
The fixup `#include` line should be together with other `#includes`. In this case, drop line 36.
Please, review same with all other soc.h files.

attie-argentum replied:
This is in the `soc/arm/atmel_sam0/...` tree, so hopefully we don't need that in the header too?

nandojve replied:
It should follow current standard file name `<driver>_fixup_<platform/soc>.h`

nandojve reviewed (changes requested):
Hi @attie-argentum ,

I think fixups makes code far more clean to read. Thank you for doing it!

> @nandojve I believe I have now implemented an adc_fixup.h as requested. I've left this as a separate patch, as it's quite far-reaching...

I will suggest that if you are willing to do, you could mode fixups before you introduce SAML21 support. This will clean the path to add the new SoC support.

attie-argentum committed:
soc: atmel_sam0: Implement fixup.h for ADC driver

The ADC driver now supports three different implementations. To maintain
readability, this patch implements an adc_fixup.h that permits more
generic access to relevant registers.

This patch also introduces support for a new third shape ADC - as found
in the SAML21 for example.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

attie-argentum commented:
@nandojve I've now re-worked the fixup header.

The SAML21 support is added transparently, as I felt this was acceptable and easiest... it won't be usable until #41308 is merged (unless another existing part has the same shape ADC).

I've also rebased on main

nandojve commented on drivers/adc/adc_sam0.c:
```diff
@@ -544,43 +521,25 @@ static const struct adc_driver_api adc_sam0_api = {
-	adc->CALIB.reg = ADC_CALIB_BIASCOMP(comp) |			\
-			 ADC_CALIB_BIASR2R(r2r) |			\
-			 ADC_CALIB_BIASREFBUF(rbuf);			\
+	adc->CALIB.reg = ADC_SAM0_BIASCOMP(n) |				\
+			 ADC_SAM0_BIASR2R(n) |				\
+			 ADC_SAM0_BIASREFBUF(n);			\
```
```c
	adc->CALIB.reg = ADC_SAM0_BIASCOMP(n)				\
		       | ADC_SAM0_BIASR2R(n)				\
		       | ADC_SAM0_BIASREFBUF(n);
```

nandojve reviewed (changes requested):
Hi @attie-argentum ,

Fixup file helped to read code. Thank you for introduce it.

attie-argentum committed:
drivers: adc: adc_sam0: Change local variables to lower case

Local variables should not be in capitals.

Signed-off-by: Attie Grande <attie.grande@argentum-systems.co.uk>

KimBP commented on drivers/adc/adc_sam0.c:
```diff
@@ -392,7 +375,20 @@ static int start_read(const struct device *dev,
+			active_channels += 1;
+		}
+	}
+	if (active_channels != 1U) {
```
I'll suggest replacing your bit check with this (more elegant and faster) (See PR 

```
-	if (sequence->channels != 1U) {
+	if ((sequence->channels == 0) || 
+		((sequence->channels & (sequence->channels - 1)) != 0)) {
```

Then you can also remove the two added variables (i and active_channels) defined 50 lines above

attie-argentum replied:
Thanks! It took me a sec to figure out what is going on, but it looks good.

attie-argentum commented:
I have some more fixes for the ADC driver coming - are is it okay for me to roll them into this PR, or would a new PR be preferred?

- REFCTRL is enable-protected on the SAML21, meaning that the ADC needs to be disabled before changing it.
    - Docs indicate that this is probably true for SAMD21 too (along with the "_discard next sample after reference change_")
    - SAME54 seems to not have this restriction
- On SAML21, the reference mapping is different
    - Zephyr's `ADC_REF_VDD_1_2` -> ASF's `ADC_REFCTRL_REFSEL_INTVCC1`
    - Zephyr's `ADC_REF_VDD_1` -> ASF's `ADC_REFCTRL_REFSEL_INTVCC2`
    - I'd propose putting this into a new per-soc fixup header.

nandojve commented:
Hi @anangl , @stephanosio , @mnkp,

Should this fix be at v3.0?

attie-argentum commented:
Rebase
